File: core/src/main/kotlin/com/example/toyracers/collision/CollisionSystem.kt
package com.example.toyracers.collision

import com.example.toyracers.car.CarState
import com.example.toyracers.track.Track
import com.example.toyracers.track.TrackCircle
import com.example.toyracers.track.TrackCollisionShape
import com.example.toyracers.track.TrackPoint
import com.example.toyracers.track.TrackPolygon
import com.example.toyracers.track.TrackRectangle
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/** Deterministic circle-based collision detection and response. */
class CollisionSystem(
    private val config: CollisionConfig = CollisionConfig(),
) {
    fun resolveTrackCollision(
        state: CarState,
        radius: Float,
        track: Track,
    ): CollisionResult {
        require(radius > 0f) { "Collision radius must be positive" }
        require(track.worldBounds.width >= radius * 2f) {
            "Collision radius does not fit inside world bounds width"
        }
        require(track.worldBounds.height >= radius * 2f) {
            "Collision radius does not fit inside world bounds height"
        }
        val contacts = mutableListOf<CollisionContact>()

        var passesRemaining = MAX_RESOLUTION_PASSES
        while (passesRemaining > 0) {
            passesRemaining--
            val contactCountBeforePass = contacts.size
            resolveOuterBoundary(state, radius, track.worldBounds, contacts)
            track.innerObstacles.forEach { obstacle ->
                resolveRectangleObstacle(state, radius, obstacle, contacts)
            }
            track.collisionShapes.forEach { shape ->
                resolveCollisionShape(state, radius, shape, contacts)
            }
            if (contacts.size == contactCountBeforePass) {
                break
            }
        }

        updateLongitudinalSpeed(state)
        return if (contacts.isEmpty()) CollisionResult.NONE else CollisionResult(contacts)
    }

    private fun resolveCollisionShape(
        state: CarState,
        radius: Float,
        shape: TrackCollisionShape,
        contacts: MutableList<CollisionContact>,
    ) {
        when (shape) {
            is TrackCircle -> resolveCircleObstacle(state, radius, shape, contacts)
            is TrackPolygon -> resolvePolygonObstacle(state, radius, shape, contacts)
        }
    }

    private fun resolveCircleObstacle(
        state: CarState,
        radius: Float,
        obstacle: TrackCircle,
        contacts: MutableList<CollisionContact>,
    ) {
        val offsetX = state.x - obstacle.center.x
        val offsetY = state.y - obstacle.center.y
        val combinedRadius = radius + obstacle.radius
        val distanceSquared = offsetX * offsetX + offsetY * offsetY
        if (distanceSquared >= combinedRadius * combinedRadius) return
        val distance = sqrt(distanceSquared)
        val normalX = if (distance > MIN_DISTANCE) offsetX / distance else 1f
        val normalY = if (distance > MIN_DISTANCE) offsetY / distance else 0f
        resolveShapeContact(
            state,
            normalX,
            normalY,
            combinedRadius - distance,
            contacts,
        )
    }

    private fun resolvePolygonObstacle(
        state: CarState,
        radius: Float,
        obstacle: TrackPolygon,
        contacts: MutableList<CollisionContact>,
    ) {
        val center = TrackPoint(state.x, state.y)
        val closest = closestPolygonEdge(center, obstacle)
        val offsetX = state.x - closest.point.x
        val offsetY = state.y - closest.point.y
        val distance = hypot(offsetX, offsetY)
        val inside = pointInsidePolygon(center, obstacle.vertices)
        if (!inside && distance >= radius) return

        val normalX: Float
        val normalY: Float
        val penetration: Float
        if (inside) {
            val outward = outwardNormal(closest.edgeStart, closest.edgeEnd, obstacle.vertices)
            normalX = outward.first
            normalY = outward.second
            penetration = radius + distance
        } else if (distance <= MIN_DISTANCE) {
            val outward = outwardNormal(closest.edgeStart, closest.edgeEnd, obstacle.vertices)
            normalX = outward.first
            normalY = outward.second
            penetration = radius
        } else {
            normalX = offsetX / distance
            normalY = offsetY / distance
            penetration = radius - distance
        }
        resolveShapeContact(state, normalX, normalY, penetration, contacts)
    }

    private fun resolveShapeContact(
        state: CarState,
        normalX: Float,
        normalY: Float,
        penetration: Float,
        contacts: MutableList<CollisionContact>,
    ) {
        state.x += normalX * penetration
        state.y += normalY * penetration
        addWallContact(
            state,
            normalX,
            normalY,
            penetration,
            CollisionType.TRACK_OBJECT,
            contacts,
        )
    }

    private fun closestPolygonEdge(
        point: TrackPoint,
        polygon: TrackPolygon,
    ): ClosestPolygonEdge {
        var closest: ClosestPolygonEdge? = null
        polygon.vertices.indices.forEach { index ->
            val start = polygon.vertices[index]
            val end = polygon.vertices[(index + 1) % polygon.vertices.size]
            val edgeX = end.x - start.x
            val edgeY = end.y - start.y
            val lengthSquared = edgeX * edgeX + edgeY * edgeY
            val fraction = (
                (point.x - start.x) * edgeX +
                    (point.y - start.y) * edgeY
                ) / lengthSquared
            val clampedFraction = fraction.coerceIn(0f, 1f)
            val edgePoint = TrackPoint(
                start.x + edgeX * clampedFraction,
                start.y + edgeY * clampedFraction,
            )
            val distance = hypot(point.x - edgePoint.x, point.y - edgePoint.y)
            if (closest == null || distance < closest!!.distance) {
                closest = ClosestPolygonEdge(edgePoint, start, end, distance)
            }
        }
        return checkNotNull(closest)
    }

    private fun pointInsidePolygon(
        point: TrackPoint,
        vertices: List<TrackPoint>,
    ): Boolean {
        var inside = false
        var previous = vertices.last()
        vertices.forEach { current ->
            if (
                (current.y > point.y) != (previous.y > point.y) &&
                point.x < (
                    (previous.x - current.x) * (point.y - current.y) /
                        (previous.y - current.y) + current.x
                    )
            ) {
                inside = !inside
            }
            previous = current
        }
        return inside
    }

    private fun outwardNormal(
        start: TrackPoint,
        end: TrackPoint,
        vertices: List<TrackPoint>,
    ): Pair<Float, Float> {
        val edgeX = end.x - start.x
        val edgeY = end.y - start.y
        val length = hypot(edgeX, edgeY)
        val signedArea = vertices.indices.sumOf { index ->
            val current = vertices[index]
            val next = vertices[(index + 1) % vertices.size]
            (current.x * next.y - next.x * current.y).toDouble()
        }
        return if (signedArea >= 0.0) {
            Pair(edgeY / length, -edgeX / length)
        } else {
            Pair(-edgeY / length, edgeX / length)
        }
    }

    private fun resolveRectangleObstacle(
        state: CarState,
        radius: Float,
        obstacle: TrackRectangle,
        contacts: MutableList<CollisionContact>,
    ) {
        val nearestX = state.x.coerceIn(obstacle.x, obstacle.maxX)
        val nearestY = state.y.coerceIn(obstacle.y, obstacle.maxY)
        val offsetX = state.x - nearestX
        val offsetY = state.y - nearestY
        val distanceSquared = offsetX * offsetX + offsetY * offsetY
        if (distanceSquared >= radius * radius) return

        val distance = sqrt(distanceSquared)
        val normalAndPenetration = if (distance > MIN_DISTANCE) {
            Triple(offsetX / distance, offsetY / distance, radius - distance)
        } else {
            insideRectangleNormal(state, radius, obstacle)
        }
        state.x += normalAndPenetration.first * normalAndPenetration.third
        state.y += normalAndPenetration.second * normalAndPenetration.third
        addWallContact(
            state = state,
            normalX = normalAndPenetration.first,
            normalY = normalAndPenetration.second,
            penetration = normalAndPenetration.third,
            type = CollisionType.TRACK_OBJECT,
            contacts = contacts,
        )
    }

    private fun insideRectangleNormal(
        state: CarState,
        radius: Float,
        obstacle: TrackRectangle,
    ): Triple<Float, Float, Float> {
        val distances = listOf(
            Triple(-1f, 0f, state.x - obstacle.x),
            Triple(1f, 0f, obstacle.maxX - state.x),
            Triple(0f, -1f, state.y - obstacle.y),
            Triple(0f, 1f, obstacle.maxY - state.y),
        )
        val nearestSide = distances.minBy { it.third }
        return Triple(nearestSide.first, nearestSide.second, radius + nearestSide.third)
    }

    fun resolveCarCollision(
        first: CarState,
        firstRadius: Float,
        second: CarState,
        secondRadius: Float,
    ): CollisionResult {
        require(firstRadius > 0f) { "First collision radius must be positive" }
        require(secondRadius > 0f) { "Second collision radius must be positive" }

        val offsetX = second.x - first.x
        val offsetY = second.y - first.y
        val distanceSquared = offsetX * offsetX + offsetY * offsetY
        val combinedRadius = firstRadius + secondRadius
        if (distanceSquared >= combinedRadius * combinedRadius) {
            return CollisionResult.NONE
        }

        val distance = sqrt(distanceSquared)
        val normalX = if (distance > MIN_DISTANCE) offsetX / distance else 1f
        val normalY = if (distance > MIN_DISTANCE) offsetY / distance else 0f
        val penetration = combinedRadius - distance
        val correction = penetration / 2f
        first.x -= normalX * correction
        first.y -= normalY * correction
        second.x += normalX * correction
        second.y += normalY * correction

        val relativeVelocityX = first.velocityX - second.velocityX
        val relativeVelocityY = first.velocityY - second.velocityY
        val closingSpeed =
            (relativeVelocityX * normalX + relativeVelocityY * normalY).coerceAtLeast(0f)
        val impulse = min(
            closingSpeed * (1f + config.carRestitution) / 2f,
            config.maxCarImpulse,
        )
        first.velocityX -= normalX * impulse
        first.velocityY -= normalY * impulse
        second.velocityX += normalX * impulse
        second.velocityY += normalY * impulse
        updateLongitudinalSpeed(first)
        updateLongitudinalSpeed(second)

        return CollisionResult(
            listOf(
                CollisionContact(
                    type = CollisionType.CAR,
                    normalX = -normalX,
                    normalY = -normalY,
                    penetration = penetration,
                    impactSpeed = closingSpeed,
                ),
            ),
        )
    }

    private fun resolveOuterBoundary(
        state: CarState,
        radius: Float,
        boundary: TrackRectangle,
        contacts: MutableList<CollisionContact>,
    ) {
        val minimumX = boundary.x + radius
        val maximumX = boundary.maxX - radius
        val minimumY = boundary.y + radius
        val maximumY = boundary.maxY - radius

        if (state.x < minimumX) {
            val penetration = minimumX - state.x
            state.x = minimumX
            addWallContact(state, 1f, 0f, penetration, CollisionType.WORLD_BOUNDARY, contacts)
        } else if (state.x > maximumX) {
            val penetration = state.x - maximumX
            state.x = maximumX
            addWallContact(state, -1f, 0f, penetration, CollisionType.WORLD_BOUNDARY, contacts)
        }

        if (state.y < minimumY) {
            val penetration = minimumY - state.y
            state.y = minimumY
            addWallContact(state, 0f, 1f, penetration, CollisionType.WORLD_BOUNDARY, contacts)
        } else if (state.y > maximumY) {
            val penetration = state.y - maximumY
            state.y = maximumY
            addWallContact(state, 0f, -1f, penetration, CollisionType.WORLD_BOUNDARY, contacts)
        }
    }

    private fun addWallContact(
        state: CarState,
        normalX: Float,
        normalY: Float,
        penetration: Float,
        type: CollisionType,
        contacts: MutableList<CollisionContact>,
    ) {
        val velocityIntoAllowedArea = state.velocityX * normalX + state.velocityY * normalY
        val impactSpeed = (-velocityIntoAllowedArea).coerceAtLeast(0f)
        if (velocityIntoAllowedArea < 0f) {
            state.velocityX -= normalX * velocityIntoAllowedArea
            state.velocityY -= normalY * velocityIntoAllowedArea
            state.velocityX *= config.wallSpeedRetention
            state.velocityY *= config.wallSpeedRetention
        }
        contacts += CollisionContact(type, normalX, normalY, penetration, impactSpeed)
    }

    private fun updateLongitudinalSpeed(state: CarState) {
        val radians = Math.toRadians(state.rotationDeg.toDouble())
        state.speed =
            state.velocityX * cos(radians).toFloat() +
            state.velocityY * sin(radians).toFloat()
        if (hypot(state.velocityX, state.velocityY) < MIN_DISTANCE) {
            state.velocityX = 0f
            state.velocityY = 0f
            state.speed = 0f
        }
    }

    private companion object {
        const val MAX_RESOLUTION_PASSES = 4
        const val MIN_DISTANCE = 0.0001f
    }

    private data class ClosestPolygonEdge(
        val point: TrackPoint,
        val edgeStart: TrackPoint,
        val edgeEnd: TrackPoint,
        val distance: Float,
    )
}

File: dart/lib/simulation/collision/collision_system.dart
import '../car/car_config.dart';
import '../car/car_state.dart';
import '../math/float32.dart';
import '../track/track.dart';
import '../track/track_geometry.dart';
import '../track/track_point.dart';
import 'collision_config.dart';
import 'collision_geometry.dart';
import 'collision_result.dart';

export 'collision_result.dart';

/// Resolves ordered contacts in the pure-Dart simulation layer.
///
/// Collision geometry and response deliberately mirror the Kotlin oracle.
/// Flame may consume the resulting state for presentation, but it must not
/// replace this simulation result.
abstract interface class CollisionSystem {
  factory CollisionSystem({CollisionConfig? config}) = DefaultCollisionSystem;

  /// Resolves a car capsule against world bounds and ordered track solids.
  CollisionResult resolveTrackCollision({
    required CarState state,
    required CarConfig config,
    required Track track,
  });

  /// Resolves the nearest pair of circles from two ordered car capsules.
  CollisionResult resolveCarCollision({
    required CarState firstState,
    required CarConfig firstConfig,
    required CarState secondState,
    required CarConfig secondConfig,
  });
}

/// Reference-compatible, bounded circle and capsule contact resolver.
final class DefaultCollisionSystem implements CollisionSystem {
  DefaultCollisionSystem({CollisionConfig? config})
    : _config = config ?? CollisionConfig();

  static const int _maxResolutionPasses = 4;

  final CollisionConfig _config;

  @override
  CollisionResult resolveTrackCollision({
    required CarState state,
    required CarConfig config,
    required Track track,
  }) {
    final circleCount = CollisionGeometry.carCircles(state, config).length;
    final contacts = <CollisionContact>[];
    for (var index = 0; index < circleCount; index++) {
      final circle = CollisionGeometry.carCircles(state, config)[index];
      contacts.addAll(_resolveTrackCircle(state, circle, track).contacts);
    }
    return _resultFor(contacts);
  }

  @override
  CollisionResult resolveCarCollision({
    required CarState firstState,
    required CarConfig firstConfig,
    required CarState secondState,
    required CarConfig secondConfig,
  }) {
    final closestPair = CollisionGeometry.closestCirclePair(
      CollisionGeometry.carCircles(firstState, firstConfig),
      CollisionGeometry.carCircles(secondState, secondConfig),
    );
    final firstCircleState = firstState.copy()
      ..x = closestPair.firstCircle.x
      ..y = closestPair.firstCircle.y;
    final secondCircleState = secondState.copy()
      ..x = closestPair.secondCircle.x
      ..y = closestPair.secondCircle.y;
    final result = _resolveCarCircles(
      firstCircleState,
      closestPair.firstCircle.radius,
      secondCircleState,
      closestPair.secondCircle.radius,
    );
    _translateState(firstState, firstCircleState, closestPair.firstCircle);
    _translateState(secondState, secondCircleState, closestPair.secondCircle);
    _copyMotion(firstState, firstCircleState);
    _copyMotion(secondState, secondCircleState);
    return result;
  }

  CollisionResult _resolveTrackCircle(
    CarState state,
    CollisionCircle circle,
    Track track,
  ) {
    final circleState = state.copy()
      ..x = circle.x
      ..y = circle.y;
    final result = _resolveTrackCircleState(circleState, circle.radius, track);
    _translateState(state, circleState, circle);
    _copyMotion(state, circleState);
    return result;
  }

  CollisionResult _resolveTrackCircleState(
    CarState state,
    double radius,
    Track track,
  ) {
    _validateRadiusFitsTrack(radius, track.worldBounds);
    final contacts = <CollisionContact>[];
    for (var pass = 0; pass < _maxResolutionPasses; pass++) {
      final contactsBeforePass = contacts.length;
      _resolveOuterBoundary(state, radius, track.worldBounds, contacts);
      for (final obstacle in track.innerObstacles) {
        _resolveRectangleObstacle(state, radius, obstacle, contacts);
      }
      for (final shape in track.collisionShapes) {
        _resolveCollisionShape(state, radius, shape, contacts);
      }
      if (contacts.length == contactsBeforePass) {
        break;
      }
    }
    CollisionGeometry.updateLongitudinalSpeed(state);
    return _resultFor(contacts);
  }

  CollisionResult _resolveCarCircles(
    CarState first,
    double firstRadius,
    CarState second,
    double secondRadius,
  ) {
    _validatePositiveRadius(
      firstRadius,
      'First collision radius must be positive',
    );
    _validatePositiveRadius(
      secondRadius,
      'Second collision radius must be positive',
    );
    final offsetX = Float32.subtract(second.x, first.x);
    final offsetY = Float32.subtract(second.y, first.y);
    final distanceSquared = CollisionGeometry.lengthSquared(offsetX, offsetY);
    final combinedRadius = Float32.add(firstRadius, secondRadius);
    if (distanceSquared >= Float32.multiply(combinedRadius, combinedRadius)) {
      return CollisionResult.none;
    }

    final distance = CollisionGeometry.distanceFromSquared(distanceSquared);
    final normal = distance > CollisionGeometry.minimumDistance
        ? CollisionNormal(
            Float32.divide(offsetX, distance),
            Float32.divide(offsetY, distance),
          )
        : const CollisionNormal(1, 0);
    final penetration = Float32.subtract(combinedRadius, distance);
    final correction = Float32.divide(penetration, 2);
    first.x = Float32.subtract(first.x, Float32.multiply(normal.x, correction));
    first.y = Float32.subtract(first.y, Float32.multiply(normal.y, correction));
    second.x = Float32.add(second.x, Float32.multiply(normal.x, correction));
    second.y = Float32.add(second.y, Float32.multiply(normal.y, correction));

    final closingSpeed = _nonNegative(
      CollisionGeometry.dot(
        Float32.subtract(first.velocityX, second.velocityX),
        Float32.subtract(first.velocityY, second.velocityY),
        normal.x,
        normal.y,
      ),
    );
    final impulse = _minimum(
      Float32.divide(
        Float32.multiply(closingSpeed, Float32.add(1, _config.carRestitution)),
        2,
      ),
      _config.maxCarImpulse,
    );
    first.velocityX = Float32.subtract(
      first.velocityX,
      Float32.multiply(normal.x, impulse),
    );
    first.velocityY = Float32.subtract(
      first.velocityY,
      Float32.multiply(normal.y, impulse),
    );
    second.velocityX = Float32.add(
      second.velocityX,
      Float32.multiply(normal.x, impulse),
    );
    second.velocityY = Float32.add(
      second.velocityY,
      Float32.multiply(normal.y, impulse),
    );
    CollisionGeometry.updateLongitudinalSpeed(first);
    CollisionGeometry.updateLongitudinalSpeed(second);
    return CollisionResult(
      contacts: <CollisionContact>[
        CollisionContact(
          type: CollisionType.car,
          normalX: Float32.narrow(-normal.x),
          normalY: Float32.narrow(-normal.y),
          penetration: penetration,
          impactSpeed: closingSpeed,
        ),
      ],
    );
  }

  void _resolveCollisionShape(
    CarState state,
    double radius,
    TrackCollisionShape shape,
    List<CollisionContact> contacts,
  ) {
    switch (shape) {
      case TrackCircle():
        _resolveCircleObstacle(state, radius, shape, contacts);
      case TrackPolygon():
        _resolvePolygonObstacle(state, radius, shape, contacts);
    }
  }

  void _resolveCircleObstacle(
    CarState state,
    double radius,
    TrackCircle obstacle,
    List<CollisionContact> contacts,
  ) {
    final offsetX = Float32.subtract(state.x, obstacle.center.x);
    final offsetY = Float32.subtract(state.y, obstacle.center.y);
    final combinedRadius = Float32.add(radius, obstacle.radius);
    final distanceSquared = CollisionGeometry.lengthSquared(offsetX, offsetY);
    if (distanceSquared >= Float32.multiply(combinedRadius, combinedRadius)) {
      return;
    }
    final distance = CollisionGeometry.distanceFromSquared(distanceSquared);
    final normal = distance > CollisionGeometry.minimumDistance
        ? CollisionNormal(
            Float32.divide(offsetX, distance),
            Float32.divide(offsetY, distance),
          )
        : const CollisionNormal(1, 0);
    _resolveShapeContact(
      state,
      normal,
      Float32.subtract(combinedRadius, distance),
      contacts,
    );
  }

  void _resolvePolygonObstacle(
    CarState state,
    double radius,
    TrackPolygon obstacle,
    List<CollisionContact> contacts,
  ) {
    final center = TrackPoint(state.x, state.y);
    final closest = CollisionGeometry.closestPolygonEdge(center, obstacle);
    final offsetX = Float32.subtract(state.x, closest.point.x);
    final offsetY = Float32.subtract(state.y, closest.point.y);
    final distance = CollisionGeometry.hypot(offsetX, offsetY);
    final inside = obstacle.contains(center.x, center.y);
    if (!inside && distance >= radius) {
      return;
    }

    final CollisionNormal normal;
    final double penetration;
    if (inside) {
      normal = CollisionGeometry.outwardNormal(
        closest.edgeStart,
        closest.edgeEnd,
        obstacle.vertices,
      );
      penetration = Float32.add(radius, distance);
    } else if (distance <= CollisionGeometry.minimumDistance) {
      normal = CollisionGeometry.outwardNormal(
        closest.edgeStart,
        closest.edgeEnd,
        obstacle.vertices,
      );
      penetration = radius;
    } else {
      normal = CollisionNormal(
        Float32.divide(offsetX, distance),
        Float32.divide(offsetY, distance),
      );
      penetration = Float32.subtract(radius, distance);
    }
    _resolveShapeContact(state, normal, penetration, contacts);
  }

  void _resolveShapeContact(
    CarState state,
    CollisionNormal normal,
    double penetration,
    List<CollisionContact> contacts,
  ) {
    state.x = Float32.add(state.x, Float32.multiply(normal.x, penetration));
    state.y = Float32.add(state.y, Float32.multiply(normal.y, penetration));
    _addWallContact(
      state,
      normal,
      penetration,
      CollisionType.trackObject,
      contacts,
    );
  }

  void _resolveRectangleObstacle(
    CarState state,
    double radius,
    TrackRectangle obstacle,
    List<CollisionContact> contacts,
  ) {
    final nearestX = Float32.clamp(state.x, obstacle.x, obstacle.maxX);
    final nearestY = Float32.clamp(state.y, obstacle.y, obstacle.maxY);
    final offsetX = Float32.subtract(state.x, nearestX);
    final offsetY = Float32.subtract(state.y, nearestY);
    final distanceSquared = CollisionGeometry.lengthSquared(offsetX, offsetY);
    if (distanceSquared >= Float32.multiply(radius, radius)) {
      return;
    }

    final distance = CollisionGeometry.distanceFromSquared(distanceSquared);
    final correction = distance > CollisionGeometry.minimumDistance
        ? CollisionPenetration(
            CollisionNormal(
              Float32.divide(offsetX, distance),
              Float32.divide(offsetY, distance),
            ),
            Float32.subtract(radius, distance),
          )
        : _insideRectangleNormal(state, radius, obstacle);
    _resolveShapeContact(
      state,
      correction.normal,
      correction.penetration,
      contacts,
    );
  }

  void _resolveOuterBoundary(
    CarState state,
    double radius,
    TrackRectangle boundary,
    List<CollisionContact> contacts,
  ) {
    final minimumX = Float32.add(boundary.x, radius);
    final maximumX = Float32.subtract(boundary.maxX, radius);
    final minimumY = Float32.add(boundary.y, radius);
    final maximumY = Float32.subtract(boundary.maxY, radius);
    if (state.x < minimumX) {
      final penetration = Float32.subtract(minimumX, state.x);
      state.x = minimumX;
      _addWallContact(
        state,
        const CollisionNormal(1, 0),
        penetration,
        CollisionType.worldBoundary,
        contacts,
      );
    } else if (state.x > maximumX) {
      final penetration = Float32.subtract(state.x, maximumX);
      state.x = maximumX;
      _addWallContact(
        state,
        const CollisionNormal(-1, 0),
        penetration,
        CollisionType.worldBoundary,
        contacts,
      );
    }
    if (state.y < minimumY) {
      final penetration = Float32.subtract(minimumY, state.y);
      state.y = minimumY;
      _addWallContact(
        state,
        const CollisionNormal(0, 1),
        penetration,
        CollisionType.worldBoundary,
        contacts,
      );
    } else if (state.y > maximumY) {
      final penetration = Float32.subtract(state.y, maximumY);
      state.y = maximumY;
      _addWallContact(
        state,
        const CollisionNormal(0, -1),
        penetration,
        CollisionType.worldBoundary,
        contacts,
      );
    }
  }

  void _addWallContact(
    CarState state,
    CollisionNormal normal,
    double penetration,
    CollisionType type,
    List<CollisionContact> contacts,
  ) {
    final velocityIntoAllowedArea = CollisionGeometry.dot(
      state.velocityX,
      state.velocityY,
      normal.x,
      normal.y,
    );
    final impactSpeed = _nonNegative(Float32.narrow(-velocityIntoAllowedArea));
    if (velocityIntoAllowedArea < 0) {
      _applyWallResponse(state, normal, velocityIntoAllowedArea);
    }
    contacts.add(
      CollisionContact(
        type: type,
        normalX: normal.x,
        normalY: normal.y,
        penetration: penetration,
        impactSpeed: impactSpeed,
      ),
    );
  }

  CollisionPenetration _insideRectangleNormal(
    CarState state,
    double radius,
    TrackRectangle obstacle,
  ) {
    final distances = <({CollisionNormal normal, double distance})>[
      (
        normal: const CollisionNormal(-1, 0),
        distance: Float32.subtract(state.x, obstacle.x),
      ),
      (
        normal: const CollisionNormal(1, 0),
        distance: Float32.subtract(obstacle.maxX, state.x),
      ),
      (
        normal: const CollisionNormal(0, -1),
        distance: Float32.subtract(state.y, obstacle.y),
      ),
      (
        normal: const CollisionNormal(0, 1),
        distance: Float32.subtract(obstacle.maxY, state.y),
      ),
    ];
    var nearest = distances.first;
    for (final side in distances.skip(1)) {
      if (side.distance < nearest.distance) {
        nearest = side;
      }
    }
    return CollisionPenetration(
      nearest.normal,
      Float32.add(radius, nearest.distance),
    );
  }

  void _applyWallResponse(
    CarState state,
    CollisionNormal normal,
    double velocityIntoAllowedArea,
  ) {
    state.velocityX = Float32.subtract(
      state.velocityX,
      Float32.multiply(normal.x, velocityIntoAllowedArea),
    );
    state.velocityY = Float32.subtract(
      state.velocityY,
      Float32.multiply(normal.y, velocityIntoAllowedArea),
    );
    state.velocityX = Float32.multiply(
      state.velocityX,
      _config.wallSpeedRetention,
    );
    state.velocityY = Float32.multiply(
      state.velocityY,
      _config.wallSpeedRetention,
    );
  }

  void _validateRadiusFitsTrack(double radius, TrackRectangle worldBounds) {
    _validatePositiveRadius(radius, 'Collision radius must be positive');
    if (worldBounds.width < Float32.multiply(radius, 2)) {
      throw ArgumentError(
        'Collision radius does not fit inside world bounds width',
      );
    }
    if (worldBounds.height < Float32.multiply(radius, 2)) {
      throw ArgumentError(
        'Collision radius does not fit inside world bounds height',
      );
    }
  }

  static void _translateState(
    CarState state,
    CarState circleState,
    CollisionCircle circle,
  ) {
    state.x = Float32.add(state.x, Float32.subtract(circleState.x, circle.x));
    state.y = Float32.add(state.y, Float32.subtract(circleState.y, circle.y));
  }

  static void _copyMotion(CarState target, CarState source) {
    target.longitudinalSpeed = source.longitudinalSpeed;
    target.velocityX = source.velocityX;
    target.velocityY = source.velocityY;
    target.lateralSpeed = source.lateralSpeed;
  }

  static void _validatePositiveRadius(double radius, String message) {
    if (radius <= 0) {
      throw ArgumentError(message);
    }
  }

  static CollisionResult _resultFor(List<CollisionContact> contacts) =>
      contacts.isEmpty
          ? CollisionResult.none
          : CollisionResult(contacts: contacts);

  static double _minimum(double left, double right) =>
      left < right ? left : right;

  static double _nonNegative(double value) => value < 0 ? 0 : value;
}
